### ISSUE TITLE
Add HashSet.AsReadOnly

### ISSUE THREAD

**Issue by vanillajonathan** [dotnet/corefx#37220]
The [`List<>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1) class have a [`AsReadOnly`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.asreadonly) method to turn the collection into a [`ReadOnlyCollection`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlycollection-1), but [`HashSet<>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1) does not have any such method.

## Rationale and usage
Allows Entity Framework Core to expose a [collection navigation property](https://docs.microsoft.com/en-us/ef/core/modeling/relationships) as a [`IReadOnlyCollection`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlycollection-1) when backing it with a private [backing field](https://docs.microsoft.com/en-us/ef/core/modeling/backing-field) declared as a `HashSet`.
```cs
[Table("Blogs")]
public class Blog
{
    private HashSet<Post> _posts;
 
    public int BlogId { get; set; }
 
    public IReadOnlyCollection<Post> Posts => _posts?.AsReadOnly();
 
    public void AddPost(Post post)
    {
        // Do some business logic here...
        _posts.Add(post);
    }
}
```

## Proposed API
```diff
public class HashSet<T> : ICollection<T>, IEnumerable<T>, IEnumerable, IReadOnlyCollection<T>, ISet<T>, IDeserializationCallback, ISerializable
{
+    public ReadOnlyCollection<T> AsReadOnly();
}

**Comment by safern** [comment]
Thanks @vanillajonathan I think this makes sense. I've marked the issue as api ready for review.

**Comment by omariom** [comment]
Returning  `ReadOnlyCollection` means `AsReadOnly` will have to make a full copy of the HashSet.

What about a new type `ReadOnlySet` than can wrap any `ISet<T>`?
Then we will have readonly wrappers for all "ObjectModel" collections: [IList](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlycollection-1?view=netcore-2.2), [IDictionary](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlydictionary-2?view=netcore-2.2), [ObservableCollection](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlyobservablecollection-1?view=netcore-2.2) and ISet.

**Comment by GrabYourPitchforks** [comment]
Feedback - what are you actually trying to do here?

`HashSet<T>` already implements `IReadOnlyCollection<T>`, so you can cast it directly: `IReadOnlyCollection<T> collection = myHashSet;`.

If you want to create a `ReadOnlyCollection<T>`, it's one line: `ReadOnlyCollection<T> collection = new ReadOnlyCollection<T>(myHashSet.ToArray());`.

**Comment by vanillajonathan** [comment]
I want to create a entity in Entity Framework Core. EF Core internally treats collections  as hash sets but they are commonly exposed as `ICollection` properties in the `DbSet`.

I intend to employ domain-driven design (DDD), so I do not want to expose it as a `ICollection` since that would allow the collection to be modified.

Since `HashSet` (as you pointed out) implements `IReadOnlyCollection<T>` I can expose it as a `IReadOnlyCollection` (instead of a `ICollection`) but since that is just exposing it as a different interface it is still possible to typecast it back to a `ICollection` and modify it.

**Comment by GrabYourPitchforks** [comment]
In your scenario, accessing the `Posts` property will return a brand new collection instance every single time. I'd instead recommend using the existing `List<T>` and `ReadOnlyCollection<T>` types, especially since it seems `HashSet<T>` might not be a good fit for the `Post` type.

Example:

```cs
public class Blog
{
    private List<Post> _posts = new List<Post>();
    private ReadOnlyCollection<Post> _postsAsReadOnlyCollection = new ReadOnlyCollection<Post>(_posts);

    public IReadOnlyCollection<Post> Posts => _postsAsReadOnlyCollection;

    public void AddPost(Post post)
    {
        // business logic here
        _posts.Add(post);
    }
}
```

Or, if you really want to use `HashSet<Post>`:

```cs
public class Blog
{
    private HashSet<Post> _posts = new HashSet<Post>();
    private ReadOnlyCollection<Post> _cachedROC = null;

    public IReadOnlyCollection<Post> Posts => _cachedROC ?? (_cachedROC = new ReadOnlyCollection<T>(_posts.ToArray()));

    public void AddPost(Post post)
    {
        // business logic
        _posts.Add(post);
        _cachedROC = null; // clear cached value if it exists
    }
}
```

**Comment by vanillajonathan** [comment]
I was under the impression that `ToList()` would create a brand new collection instance every single time, and that `AsReadOnly` would not do so.

In the first of your examples I would need two private and one public instead of just one public or one private setter. Also it uses List while EF Core internally uses HashSet so the performance may differ.

In the second of your examples, it introduces caching into the business entity.

Both of these are undesirable. Could `HashSet` have a `AsReadOnly` method, and would that solve this in a clean, efficient manner?

**Comment by GrabYourPitchforks** [comment]
> I was under the impression that ToList() would create a brand new collection instance every single time, and that AsReadOnly would not do so.

Yes, it must create a brand new collection every time. The reason for this is that `ReadOnlyCollection<T>` requires the underlying collection to implement `IList<T>`, which `HashSet<T>` does not do on its own. This means that a `AsReadOnly()` method which returns a `ReadOnlyCollection<T>` necessarily needs to make a copy of the contents of the `HashSet<T>` instance.

**Comment by vanillajonathan** [comment]
> The reason for this is that `ReadOnlyCollection<T>` requires the underlying collection to implement `IList<T>`, which `HashSet<T>` does not do on its own.

This raises the question, should `HashSet<T>` implement `IList<T>`?

**Comment by GrabYourPitchforks** [comment]
> This raises the question, should HashSet<T> implement IList<T>?

No. `HashSet<T>` is an unordered no-duplicate collection. `IList<T>` is an ordered list, potentially with duplicates. They represent two different collection concepts.

**Comment by vanillajonathan** [comment]
> > This raises the question, should HashSet implement IList?
> 
> No. `HashSet<T>` is an unordered no-duplicate collection. `IList<T>` is an ordered list, potentially with duplicates. They represent two different collection concepts.

Good point. Could  `HashSet<T>` have a `AsReadOnly` method (like `List<T>` does, but without implementing `IList<T>`)?

With a `List<T>` you can cleanly do things like:
```cs
public class Blog
{
    private List<Post> _posts;
    public IReadOnlyCollection<Post> Posts => _posts?.AsReadOnly();
}
```

Which you cannot do as cleanly when using a `HashSet<T>` instead of a `List<T>`.

**Comment by GrabYourPitchforks** [comment]
Which brings us back to the original response on the thread... given your scenario it really would just be easier to have a property that directly casts the `HashSet<T>` to an `IReadOnlyCollection<T>`.

> Since HashSet (as you pointed out) implements IReadOnlyCollection<T> I can expose it as a IReadOnlyCollection (instead of a ICollection) but since that is just exposing it as a different interface it is still possible to typecast it back to a ICollection and modify it.

So don't cast it back? :)

**Comment by vanillajonathan** [comment]
> 
> 
> Which brings us back to the original response on the thread... given your scenario it really would just be easier to have a property that directly casts the `HashSet<T>` to an `IReadOnlyCollection<T>`.
> 
> > Since HashSet (as you pointed out) implements IReadOnlyCollection I can expose it as a IReadOnlyCollection (instead of a ICollection) but since that is just exposing it as a different interface it is still possible to typecast it back to a ICollection and modify it.
> 
> So don't cast it back? :)

I wouldn't typecast it back to a `ICollection<T>`, but if I exposed a `IReadOnlyCollection<T>` then I want that to be read-only, but someone using my library _could_ typecast it back a `ICollection<T>` and modify it.
With a `List<T>` this is much easier to prevent since it has a `AsReadOnly` method.

**Comment by GrabYourPitchforks** [comment]
So what is your concrete proposal then? Are you proposing a new type `ReadOnlyCollectionWrapper<T>` whose ctor takes an `IReadOnlyCollection<T>` instead of an `IList<T>`? Is this a scenario that is common enough and that would be burdensome enough for somebody to do themselves that such a type belongs in the Framework proper? Presumably if you need to be unblocked immediately you've already created such a wrapper.

**Comment by vanillajonathan** [comment]
Well I don't know how feasible it is, but I expected the `HashSet<T>` to have a `AsReadOnly` method, so I would propose the addition of a `AsReadOnly` method to the `HashSet<T>`.

**Comment by GrabYourPitchforks** [comment]
And what would be the return type of `HashSet<T>.AsReadOnly()`? As discussed it can't be `ReadOnlyCollection<T>`.

**Comment by vanillajonathan** [comment]
> And what would be the return type of `HashSet<T>.AsReadOnly()`? As discussed it can't be `ReadOnlyCollection<T>`.

You're right, it cannot return a `ReadOnlyCollection<T>` since that class implements `IList<T>` which `HashSet<T>` does not.

Maybe it is weird that `ReadOnlyCollection<T>` implements `IList<T>`, and that `IList<T>` implements `ICollection<T>`?
Maybe `ReadOnlyCollection<T>` should not have implemented `IList<T>`?

**Comment by GrabYourPitchforks** [comment]
Yeah, in hindsight it may have made sense to have two types `ReadOnlyList<T>` and `ReadOnlyCollection<T>`. That predates my time on this team so I'm not familiar with the history behind why this decision was made. But regardless it's not a decision that can be undone right now. ☹

**Comment by vanillajonathan** [comment]
Maybe things can be fixed for .NET 5?

This would be nice for developers using Entity Framework Core and who wish to implement domain-driven design (DDD).

**Comment by GrabYourPitchforks** [comment]
That's not the type of breaking change we would take in .NET 5.
